Documentation/EmacsConfiguration.md

```diff
@@ -0,0 +1,61 @@
+# Emacs Project Configuration
+
+Emacs can be configured with `lsp-mode` and `clangd` to work well.
+
+### clangd
+
+The official clangd extension can be used for C++ comprehension. You
+can use the following `.clangd` file placed in the project root:
+
+```yaml
+CompileFlags:
+  CompilationDatabase: Build/i686
+  Add:
+    - "-D__serenity__"
+    - "-UNO_TLS"
+    - "-I/path/to/serenity/Toolchain/Local/i686/i686-pc-serenity/include/c++/11.2.0"
+    - "-I/path/to/serenity/Toolchain/Local/i686/i686-pc-serenity/include/c++/11.2.0/i686-pc-serenity"
+```
+
+You will obviously need to update `11.2.0` to whatever your version at
+the time is.
+
+Run cmake at least once for this to work. clangd has difficulty
+finding specific methods and types, especially with inheritance
+trees. Also, include errors are wrong in 90% of cases.
+
+### lsp-mode
+
+```lisp
+(use-package lsp-mode
+  :hook ((c++-mode) . lsp-deferred)
+  :commands lsp
+  :config
+  (setq lsp-clients-clangd-args '("-j=4" "-background-index" "--log=error" "--clang-tidy" "--enable-config"))
+  (setq lsp-clangd-binary-path "/usr/bin/clangd"))
+
+```
+
+You can change the binary path to point to your local copy as shown,
+or your Serenity copy if you built the clang toolchain. Additionally,
+if you didn't specify these, `lsp-mode` may offer to install clang for
+you which would end up in `~/.emacs.d/cache/...` by default.
+
+### clang-format
+
+You can also install `clang-format.el` to call `clang-format` on a
+buffer or region. It might be useful to configure `clang-format` to be
+called every time you save a C/C++ buffer.
+
+```lisp
+;; Install clang-format.el.
+(use-package clang-format
+  :config
+  (setq clang-format-executable "/usr/bin/clang-format"))
+
+;; Optionally format the buffer when you save the current file.
+(add-hook 'c-mode-common-hook
+          (lambda ()
+            (add-hook 'before-save-hook 'clang-format-buffer nil 'make-it-local)))
+
+```
```

Comment:
I use [format-all-mode](https://github.com/lassik/emacs-format-all-the-code) which works pretty well with `clang-format` and also works with many other major modes. Is there a specific reason to use the `clang-format` package?

Reply:
Only because the name seemed the most generic/official. I hadn't heard of `format-all-mode` before. Since it touches things outside of Serenity/C++ it seems a bit more intrusive too, but probably not. 

What about just removing that part and just saying something like

> For formatting use a package of your choice, here are some options:
> -  format-all-mode (linked)
> - clang-format (linked)

If you have a simple config for `format-all-mode` I'm happy to just update it with that config and kill the `clang-format` stuff I did. I don't feel strongly about this part at all, so I'm cool with changing it completely.

Reply:
Yeah, that would be a lot nicer since it doesn't give a preference to a specific package.

> If you have a simple config for format-all-mode I'm happy to just update it with that config

The great part of `format-all-mode` is that it doesn't need a config! :^) It should just pick up the `clang-format` that's installed. Although that might cause some troubles since it's not easily configurable and the project now requires `clang-format` 14 that might not be system wide installed.

Reply:
Thanks for the recommendation, I'll check it out. I'm currently using separate ones for `go`, `rust`, `c++`, etc. Would be nice to have a single one.

I updated the Serenity docs just list the options as you said.